### PR TITLE
Recognition progress in C API

### PR DESCRIPTION
**Pull request by j-kubik**
I need to use progress reporting while using C API. I have written bindings for the functionality.

### CODE REVIEW

**Comment by zdenop** [comment]
AppVeyor faild with error:
[00:05:45]   c:\projects\tesseract\src\ccutil\ocrclass.h(179): error C4716: 'ETEXT_DESC::default_progress_func': must return a value

**Comment by zdenop** [comment]
Thanks. Do you have a test case for it?

**Comment by j-kubik** [comment]
@zdenop I didn't find any test for the progress reporting at all so I didn't bother. I could add some test cases if required.

**Comment by zdenop** [comment]
1. You are modifying API, so it would be great to see why do you propose change.
2. There is not much contributors for C-API 
3. Generally: it would be great to have test suit for testing tesseract commit quality (e.g. if future commits break some tesseract usage, it should be identified ASAP)

**Comment by j-kubik** [comment]
@stweil The typo has been corrected.

**Comment by j-kubik** [comment]
@zdenop 
Ad 1. The change to the API was proposed (as described in commit message) because the progress reporting callback has no 'this' parameter, and is therefore forced to use global variables if some additional data/objects need to be accessed. This is visible even in the test case I created. I had to create a static `instance` pointer in order to work-around that limitation.
Ad 2. I need the C-API so I wrote functions I needed. Just thought it could be checked into the main repo in case anybody else wanted to use these as well.
Ad 3. I added tests for the progress reporting API.

**Comment by zdenop** [comment]
thanks

File: unittest/progress_test.cc [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    ETEXT_DESC monitor;
+
+    ClassicMockProgressSink()
+    {
+        monitor.progress_callback = []( int progress, int, int, int, int ) ->bool {
+          return instance->classicProgress( progress );
+        };
+        monitor.cancel = []( void* ths, int words ) -> bool {
+          ((ClassicMockProgressSink*)ths)->cancel( words );
+        };
```
Comment (stweil) [inline unittest/progress_test.cc]: Here the compiler issues a warning:

    ../../../../unittest/progress_test.cc: In Lambda-Funktion:
    ../../../../unittest/progress_test.cc:59:9: Warnung: keine return-Anweisung in nicht void zurückgebender Funktion [-Wreturn-type]
             };
             ^

It is also reported by Coverity Scan ("CID 1393240 Missing return statement").

**Comment by zdenop** [comment]
@j-kubik : Can you have a look at #1622 ?

**Comment by wpapp** [comment]
it's ok now,thank u.
@stweil

**Pull request merged**